File: python-algorithm/leetcode/problem_1642.py
"""1642. Furthest Building You Can Reach
https://leetcode.com/problems/furthest-building-you-can-reach/
"""
from typing import List
import heapq


class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        hq = []
        for i in range(len(heights) - 1):
            diff = heights[i + 1] - heights[i]
            if diff > 0:
                heapq.heappush(hq, diff)
            if len(hq) > ladders:
                bricks -= heapq.heappop(hq)
            if bricks < 0:
                return i
        return len(heights) - 1

File: java-algo/src/main/java/com/leetcode/Problem1642.java
package com.leetcode;

import java.util.PriorityQueue;

/**
 * 1642. Furthest Building You Can Reach
 * https://leetcode.com/problems/furthest-building-you-can-reach/
 */
public class Problem1642 {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        int n = heights.length;
        int total = 0;
        PriorityQueue<Integer> q = new PriorityQueue<>();
        for (int i = 1; i < n; i++) {
            int h = heights[i] - heights[i - 1];
            if (h > 0) {
                q.offer(h);
                if (q.size() > ladders) {
                    total += q.poll();
                }
                if (total > bricks) {
                    return i - 1;
                }
            }
        }
        return n - 1;
    }
}
